### PR TITLE
Forbid `-` in namespace in reroute processor

### LINKED ISSUE #112936: reroute processor sanitization inconsistent with spec
### Elasticsearch Version

8.15

### Installed Plugins

_No response_

### Java Version

_bundled_

### OS Version

Linux carson-elastic 6.8.0-44-generic #44-Ubuntu SMP PREEMPT_DYNAMIC Tue Aug 13 13:35:26 UTC 2024 x86_64 x86_64 x86_64 GNU/Linux

### Problem Description

- `reroute` processor allows `-` in namespace while it is said to be forbidden in [docs](https://www.elastic.co/guide/en/ecs/current/ecs-data_stream.html#_data_stream_field_details).
- [substring](https://github.com/elastic/elasticsearch/blob/33d63fa9c2fc3afb9184f31e2681942cfd8dc7fd/modules/ingest-common/src/main/java/org/elasticsearch/ingest/common/RerouteProcessor.java#L227) usage in code gets the first 100 chars (as in bytes, not in unicode chars /unicode code points). This may result in invalid unicode after sanitization. Clarify whether "100 characters" in doc refers to unicode characters. If so, the code needs to be updated to be consistent with doc.

### Steps to Reproduce

```
PUT _index_template/logs-foo
{
  "data_stream": {},
  "index_patterns": ["logs-foo-*"], 
  "composed_of": ["logs-foo@mappings"],
  "ignore_missing_component_templates": ["logs-foo@mappings"],
  "template": {
    "settings": {
      "index.default_pipeline": "logs-foo@pipeline"
    }
  }, 
  "priority": 250
}

PUT _component_template/logs-foo@mappings
{
 "template": {
   "mappings": {
      "properties": {
        "@timestamp": {
          "type": "date"
        }
      }
    }
 } 
}

PUT _ingest/pipeline/logs-foo@pipeline
{
  "processors": [
    {"reroute": {
      "namespace": ["my-new-namespace"]
    }}
  ]
}

POST logs-foo-default/_doc
{
  "@timestamp": "1970-01-01T00:00:00.000Z"
}

GET logs-foo-*/_search
```

Resulting doc:
```json
{
  "took": 1,
  "timed_out": false,
  "_shards": {
    "total": 1,
    "successful": 1,
    "skipped": 0,
    "failed": 0
  },
  "hits": {
    "total": {
      "value": 1,
      "relation": "eq"
    },
    "max_score": 1,
    "hits": [
      {
        "_index": ".ds-logs-foo-my-new-namespace-2024.09.16-000001",
        "_id": "kfs8-5EB0otqeM1lzk3-",
        "_score": 1,
        "_source": {
          "data_stream": {
            "namespace": "my-new-namespace",
            "type": "logs",
            "dataset": "foo"
          },
          "@timestamp": "1970-01-01T00:00:00.000Z"
        }
      }
    ]
  }
}
```

### Logs (if relevant)

_No response_

### PR DESCRIPTION
**Pull request by carsonip**
`-` character is not allowed in ECS data_stream.namespace spec.

Fixes #112936 

- Have you signed the [contributor license agreement](https://www.elastic.co/contributor-agreement)?
- Have you followed the [contributor guidelines](https://github.com/elastic/elasticsearch/blob/main/CONTRIBUTING.md)?
- If submitting code, have you built your formula locally prior to submission with `gradle check`?
- If submitting code, is your pull request against main? Unless there is a good reason otherwise, we prefer pull requests against main and will backport as needed.
- If submitting code, have you checked that your submission is for an [OS and architecture that we support](https://www.elastic.co/support/matrix#show_os)?
- If you are submitting this code for a class then read our [policy](https://github.com/elastic/elasticsearch/blob/main/CONTRIBUTING.md#contributing-as-part-of-a-class) for that.

### CODE REVIEW

File: docs/changelog/113218.yaml [hunk 1]
```diff
@@ -0,0 +1,6 @@
+pr: 113218
+summary: Forbid - in namespace in reroute processor
+area: Ingest Node
+type: bug
```
Comment (carsonip) [inline docs/changelog/113218.yaml:4]: Not sure if "bug" is the best description. This is going to break use cases where people are using non-compliant namespace with `-` in it.

**Comment by mattc58** [comment]
Thanks @carsonip. Wouldn't this be a breaking change though? If a user has successfully used this in pipelines to create namespaces with the `-` character wouldn't this now break them?

**Comment by carsonip** [comment]
> Thanks @carsonip. Wouldn't this be a breaking change though? If a user has successfully used this in pipelines to create namespaces with the - character wouldn't this now break them?

Correct. That is also why I left the above comment about breaking use cases where people rely on the bug.

There are 2 parts to this change, one part at factory and another at runtime. The factory part does worry me a bit as valid processor will become invalid on upgrade. I don't know whether it will just stop ingestion or make ES fail to start. Now that I think more about the implications, I'm more inclined to ship it in 9.0 with no backports.

On the apm-server end, we are likely going to ship it as a bugfix in 8.16 where we change from no sanitization to replacing all invalid chars, but the impact surface will be limited to use cases where user specifically configured data stream routing attributes in OpenTelemetry (see https://github.com/elastic/apm-server/issues/14043), and it will just index documents to a different data stream at max.

File: docs/changelog/113218.yaml (hunk 1 above)
Comment (carsonip) [reply docs/changelog/113218.yaml:4]: @mattc58 should it be something else other than "bug"?

**Comment by carsonip** [comment]
Would be good to get @felixbarny 's eyes on this as it is initially written explicitly in a way that allows `-` in namespace but not dataset.

**Comment by carsonip** [comment]
Closing this PR as we may approach the issue https://github.com/elastic/elasticsearch/issues/112936 with an opposite fix.

**Pull request closed without merging**